### PR TITLE
GRPC instrumentation implementation

### PR DESCRIPTION
**Pull request by nguyenquangminh0711**
Hello there, my name is Minh Nguyen, from Employment Hero. I have a talk with Peter in Instana support center about the gRPC support. This pull is the implementation of gRPC support for the Ruby Sensor. As gRPC is official supported by Google and a official library is provided (https://github.com/grpc/grpc/tree/master/src/ruby), I follow the implementation of this library. I already test in my Company's environment and everything works like charm, except some problems with front-end display because `rpc-client` and `rpc-server` span kinds are not supported yet. So in this implementation, I treat `rpc-client` and `rpc-server` as a custom span kind. Perhaps after you guys add the support for them, I'll add to the registered span list.

Please give me some feedbacks about this implementation. Hope that the code quality is not so bad 😸 

cc @tienle, @longseespace

### CODE REVIEW

File: lib/instana/config.rb [hunk 1]
```diff
@@ -39,6 +39,7 @@ def initialize
       @config[:excon]              = { :enabled => true }
       @config[:nethttp]            = { :enabled => true }
       @config[:'rest-client']      = { :enabled => true }
+      @config[:'grpc']             = { :enabled => true }
```
Comment (pglombardo) [inline lib/instana/config.rb]: I think we can change this globally to just `:grpc`.  The quotes were only needed when there is a dash or underscore in the name.

File: lib/instana/instrumentation/grpc.rb [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+        context = ::Instana.tracer.context
+        if context
+          options[:metadata] = (options[:metadata] || {}).merge(
+            'x-instana-t' => context.trace_id_header,
+            'x-instana-s' => context.span_id_header
+          )
+        end
+
+        #{call_type}_without_instana(method, *others, **options)
```
Comment (pglombardo) [inline lib/instana/instrumentation/grpc.rb]: Was this intentionally commented out or debug remnant?

Comment (nguyenquangminh0711) [inline lib/instana/instrumentation/grpc.rb]: This is a bug from github rendering. This line is in the `class_eval` scope and call_type is interpolated into the code snippet for evaluation.

**Comment by pglombardo** [comment]
This is an excellent PR @nguyenquangminh0711 - thanks!

I've added a couple code comments/questions.

The Travis tests seem to be failing because of [this grpc issue](https://github.com/grpc/grpc/issues/11346).  Locally on OSX grpc tests pass but the ruby process segfaults on exit.  I'm looking into this.

So all minor stuff.  I'll do some more testing locally and post back here.

File: lib/instana/instrumentation/grpc.rb (hunk 2 above)
Comment (pglombardo) [inline lib/instana/instrumentation/grpc.rb]: Ah that's right.  I saw the pound sign and got tricked.  Makes sense.

**Comment by pglombardo** [comment]
Tested this instrumentation and it looks great.  We're just waiting for the next Instana release which will add UI and backend support for the new rpc span types.  Next release is schedule for the middle of next week.

**Comment by pglombardo** [comment]
GRPC span type support will be going out tomorrow.  We're also waiting on [this GRPC release](https://github.com/grpc/grpc/issues/11346#issuecomment-309188836) to fix the Travis test failures.

**Comment by nguyenquangminh0711** [comment]
That's great 👍 . The next release for  next gRPC version is scheduled in this month. Hope to see that soon 😸

**Comment by pglombardo** [comment]
The good news is that grpc 1.4 is out and tests pass 🎉 but with grpc 1.4, [the process now aborts on exit](https://travis-ci.org/instana/ruby-sensor/jobs/241951682) with `assertion failed: cqd->completed_head.next == (uintptr_t)&cqd->completed_head`.

**Comment by nguyenquangminh0711** [comment]
Hi, @pglombardo, I updated the gRPC server starting process and all the tests passed now. The problem is that we start the gRPC server in a thread different from the main thread. I think gRPC doesn't handle this case well. So, after the process stop and the gRPC thread is killed, gRPC server is not cleaned up carefully and abort with 134 status. In the patch, I stop the server explicitly after the Minitest runs. I believe that this pull could be closed now 💃.

P/s: sorry for spamming your Travis CI builds. I could not reproduce on my local machine nor Docker, so I had no choice beside testing directly on Travis CI

**Comment by pglombardo** [comment]
Excellent @nguyenquangminh0711!  Thanks for resolving that.  Going to merge, test locally and create a release soon.

**Comment by pglombardo** [comment]
This has been released in gem 1.5.0:
https://github.com/instana/ruby-sensor/releases/tag/1.5.0
https://twitter.com/pglombardo/status/882529416869556225

**Pull request merged**